tests/test_fieldset.py

```diff
@@ -931,10 +970,16 @@ def generate_dataset(xdim, ydim, zdim=1, tdim=1):
     pset = pset_type[pset_mode]['pset'](fieldset, JITParticle, 0, 0, depth=20)
 
     pset.execute(AdvectionRK4, dt=1, runtime=10)
+
+    p0 = None
+    if pset_mode in ['soa', 'aos']:
+        p0 = pset[0]
+    elif pset_mode == 'nodes':
+        p0 = pset.begin().data
```

Comment:
`pset[0]` is not guaranteed to be `pset.begin().data` because that index in the parantheses has no real meaning in nodes, because they are sorted according to the ID, and not chronologically for the time the particle was added. Because the ID generation (apart from a simple sequential generator) is based on the geospatial coordinates and because of ID recovery and particle removal, the first particle added to `pset` is clearly `pset[0]` in indexable collections, but could be anywhere (i.e. at any index) in a list collection. Real lists (i.e. not the array-like structure of what python names `list`) just are different from arrays (or: indexable collections), and indexing real lists just makes no sense, sorry.

In short: one can write `pset[0]` for nodes too, but it's rarely gonna return what you expect and (with node removal) one could even get an invalid object. Apart from that, in real work (not tests), going through indices of node-based lists is super-slow. Non-indexable collections are used differently, as you've already seen. Your suggestion with `pset[0]` already exists, but by defintion does not result in `pset.begin().data` (see reason above). `pset[-1]` may also work - that depends on the `sortedcollections` behaviour. Question is: what would you expect `pset[3]` to evaluate for ?


This very line (l. 978), and this one very test the comment is done though, is an exception where `pset[0]` for nodes would evaluate to the same as in other collections.